src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private static final double INCREASING_COEFFICIENT = 1.5;
+
+    private T[] elementData = (T[]) new Object[DEFAULT_SIZE];
+    private int size;
 
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow(elementData.length);
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index passed to the method is invalid");
+        }
+        if (DEFAULT_SIZE > elementData.length || size < elementData.length) {
+            System.arraycopy(elementData, index, elementData,
+                    index + 1, size - index);
+            elementData[index] = value;
+        } else {
+            T[] widerList = grow(elementData.length);
+            System.arraycopy(elementData, index, widerList, index + 1, size - index);
+            widerList[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index passed to the method is invalid");
+        }
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index passed to the method is invalid");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index passed to the method is invalid");
+        }
+        return droppedValue(elementData, index);
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        for (int i = 0; i < size; i++) {
+            if (element == elementData[i] || element.equals(elementData[i])) {
```

Comment:
what if `element` is null and `elementData[i]` is not null?

Reply:
if elment == nul and elementData[i] != null, then in first part of condition will "false". This means that in this case, this loop iteration will be skipped. And this is exactly what we need! Or am I wrong?

Reply:
if elment == nul and elementData[i] != null, then in first part of condition will "false". This means that in this case, this loop iteration will be skipped. And this is exactly what we need! Or am I wrong?